src/coreComponents/constitutive/unitTests/testCompositionalProperties.cpp

```diff
@@ -0,0 +1,536 @@
+std::vector< CompositionalPropertiesTestData< NC > > generateTestData();
+
+template< integer NC >
+std::unique_ptr< TestFluid< NC > > createFluid();
+template<>
+std::unique_ptr< TestFluid< 4 > > createFluid< 4 >()
+{
+  return TestFluid< 4 >::create( {Fluid::N2, Fluid::C8, Fluid::C10, Fluid::H2O} );
+}
+
+template< typename EOS_TYPE, integer NC >
+class CompositionalPropertiesTestDataTestFixture : public ::testing::TestWithParam< CompositionalPropertiesTestData< NC > >
+{
+public:
+  static constexpr integer numComps = NC;
+public:
+  CompositionalPropertiesTestDataTestFixture()
+    : m_fluid( createFluid< NC >() )
+  {}
+  ~CompositionalPropertiesTestDataTestFixture() = default;
+
+  // Compares the calculated molar density against the expected value from PVT package
+  void testMolarDensity( CompositionalPropertiesTestData< NC > const & data )
+  {
+    real64 const pressure = std::get< 0 >( data );
+    real64 const temperature = std::get< 1 >( data );
+    array1d< real64 > composition;
+    TestFluid< NC >::createArray( composition, std::get< 2 >( data ));
+    real64 const expectedMolarDensity = std::get< 3 >( data );
+
+    real64 molarDensity = 0.0;
+    computeMolarDensity( pressure,
+                         temperature,
+                         composition,
+                         molarDensity );
+    checkRelativeError( molarDensity, expectedMolarDensity, relTol, absTol );
+  }
+
+  // Compares the calculated molar density derivatives against numerical calculated
+  // finite difference values
+  void testMolarDensityDerivative( CompositionalPropertiesTestData< NC > const & data )
+  {
+    real64 const pressure = std::get< 0 >( data );
+    real64 const temperature = std::get< 1 >( data );
+    array1d< real64 > composition;
+    TestFluid< NC >::createArray( composition, std::get< 2 >( data ));
+
+    real64 molarDensity = 0.0;
+    real64 currentMolarDensity = 0.0;
+    real64 fdDerivative = 0.0;
+    real64 dMolarDensity_dp = 0.0;
+    real64 dMolarDensity_dt = 0.0;
+    array1d< real64 > dMolarDensity_dz( numComps );
+
+    computeMolarDensity( pressure,
+                         temperature,
+                         composition,
+                         molarDensity,
+                         dMolarDensity_dp,
+                         dMolarDensity_dt,
+                         dMolarDensity_dz );
+
+    // Compare against numerical derivatives
+    // -- Pressure derivative
+    real64 const dp = 1.0e-4 * pressure;
+    computeMolarDensity( pressure-dp, temperature, composition, currentMolarDensity );
+    fdDerivative = -(currentMolarDensity - molarDensity) / dp;
+    checkDerivative( dMolarDensity_dp, fdDerivative, "Molar density left pressure derivative" );
+    computeMolarDensity( pressure+dp, temperature, composition, currentMolarDensity );
+    fdDerivative = (currentMolarDensity - molarDensity) / dp;
+    checkDerivative( dMolarDensity_dp, fdDerivative, "Molar density right pressure derivative" );
```

Comment:
Optionally we could also refactor these left and right test for all _constitutive/unitTests_, it would help readability

Reply:
This is a good idea. I will do a follow up PR to address this across all the unit tests that I have added that do this type of numerical derivative test.